cmd/osbuild-jobsite-builder/main.go

```diff
@@ -206,6 +209,46 @@ func (builder *Builder) HandlePopulate(w http.ResponseWriter, r *http.Request) e
 	if r.Method != "POST" {
 		return fmt.Errorf("Builder.HandlePopulate: unexpected request method")
 	}
+	storePath := path.Join(argBuildPath, "store")
+	err := os.Mkdir(storePath, 0755)
+	if err != nil {
+		return fmt.Errorf("Builder.HandlePopulate: failed to make store directory: %v", err)
+	}
+
+	tarReader := tar.NewReader(r.Body)
+	for header, err := tarReader.Next(); err != io.EOF; header, err = tarReader.Next() {
+		if err != nil {
+			return fmt.Errorf("Builder.HandlerPopulate: failed to unpack sources: %v", err)
+		}
+
+		if filepath.Clean(header.Name) != strings.TrimSuffix(header.Name, "/") {
+			return fmt.Errorf("Builder.HandlerPopulate: name not clean: %v != %v", filepath.Clean(header.Name), header.Name)
+		}
+
+		switch header.Typeflag {
+		case tar.TypeDir:
+			if err := os.Mkdir(path.Join(storePath, header.Name), header.FileInfo().Mode()); err != nil {
+				return fmt.Errorf("Builder.HandlerPopulate: unable to make dir in sources: %v", err)
+			}
+		case tar.TypeReg:
+			err = os.MkdirAll(filepath.Dir(path.Join(storePath, header.Name)), header.FileInfo().Mode())
```

Comment:
I'm not sure we need this, technically yes but I think we can assume that tar gives us a "well" behaviring tarfile. What I don't like about it is that we now need to make assumptions about dir permissions. Using the permissions of the file (as it's done here) is probably not giving us what we want. I personally would just drop this and mandate that tar files contain the directories for the files :)